zonemaster/zonemaster-engine
# Refactor ASNLookup code and documentation

## Purpose

This PR is four-fold: 
- It updates the ASN Lookup function of Zonemaster by using standard recursive DNS lookups instead of expecting dedicated zones.
- It fixes the backup source functionality (see https://github.com/zonemaster/zonemaster-engine/pull/1311#issuecomment-1845380408)
- It removes documentation and code for deprecated profile parameter `asnroots`
- It updates documentation and unitary tests for profile parameters `asn_db.style` and `asn_db.sources`

## Context

Fixes #1255, #1052, #1053, #835

Also follows up on https://github.com/zonemaster/zonemaster-engine/pull/1311#issuecomment-1845380408

Requires #1288

## Changes

- Update code logic in `Zonemaster::Engine::ASNLookup->_cymru_asn_lookup()`
- Add message tag `ASN_LOOKUP_SOURCE` (with `level = DEBUG` , `module = System` and `testcase = Connectivity{03,04}`)
- Fix backup source functionality in `Zonemaster::Engine::ASNLookup`
- Remove `asnroots` profile property
- Refactoring
- Update documentation
- Add and update unitary tests

## How to test this PR

Tests should pass.

**Manual testing:**
1. Add ASNLookup servers `fake.asnlookup.zonemaster.net` and `asnlookup.dufberg.se` in the default profile (or just create a new one):
```
diff --git a/share/profile.json b/share/profile.json
index 2340c9fa..5ee8dc17 100644
  --- a/share/profile.json
  +++ b/share/profile.json
@@ -2,7 +2,7 @@
     "asn_db" : {
         "style" : "Cymru",
         "sources" : {
 -            "Cymru" : [ "asnlookup.zonemaster.net", "asn.cymru.com" ],
 +            "Cymru" : [ "fake.asnlookup.zonemaster.net", "asnlookup.dufberg.se" ],
             "RIPE" : [ "riswhois.ripe.net" ]
         }
     },
```
2. Start a test for Connectivity03 and Connectivity04: they should be able to reach the working ASN database server (asnlookup.dufberg.se), after skipping the non-working one (fake.asnlookup.zonemaster.net):
```
$ zonemaster-cli --show-module --show-testcase --test connectivity03 --test connectivity04 zonemaster.net --raw --level DEBUG | egrep -i "ASN_LOOKUP_SOURCE|_DIFFERENT_"
   1.11 DEBUG    System       Connectivity03 ASN_LOOKUP_SOURCE  name="fake.asnlookup.zonemaster.net"
  11.25 DEBUG    System       Connectivity03 ASN_LOOKUP_SOURCE  name="asnlookup.dufberg.se"
[ ... ]
  20.14 INFO     Connectivity Connectivity03 IPV4_DIFFERENT_ASN  asn_list=2486,8674,56908
  20.14 INFO     Connectivity Connectivity03 IPV6_DIFFERENT_ASN  asn_list=2486,8674,56908
   0.08 DEBUG    System       Connectivity04 ASN_LOOKUP_SOURCE  name="fake.asnlookup.zonemaster.net"
   0.08 DEBUG    System       Connectivity04 ASN_LOOKUP_SOURCE  name="asnlookup.dufberg.se"
[ ... ]
   0.09 INFO     Connectivity Connectivity04 CN04_IPV4_DIFFERENT_PREFIX  ns_list=ns2.nic.fr/192.93.0.4;nsa.dnsnode.net/194.58.192.46;nsp.dnsnode.net/194.58.198.32;nsu.dnsnode.net/185.42.137.98
   0.09 INFO     Connectivity Connectivity04 CN04_IPV6_DIFFERENT_PREFIX  ns_list=ns2.nic.fr/2001:660:3005:1::1:2;nsa.dnsnode.net/2a01:3f1:46::53;nsp.dnsnode.net/2a01:3f1:3032::53;nsu.dnsnode.net/2a01:3f0:400::32
```
